### ISSUE TITLE
Installation issue: clingo-bootstrap: Could NOT find Python

### ISSUE THREAD

**Issue by lrtfm** [LLNL/spack#31717]
### Steps to reproduce the issue

Hi, @alalazo  @tgamblin,  This is in the bootstrap stage. Could you help find the reason?

### Error message

<details><summary>Error message</summary><pre>
[lrtfm@ln0%tianhe2 spack-src]$ spack install tmux
==> Warning: Spec clingo-bootstrap@spack%gcc@9.3.0~docs~ipo+python build_type=Release arch=linux-rhel5-x86_64 ^python@3.6%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-rhel5-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@9.3.0~docs~ipo+python build_type=Release arch=linux-rhel5-x86_64 ^python@3.6%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-rhel5-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64 ^python@3.6%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64 ^python@3.6%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-x86_64 did not provide module clingo
==> Bootstrapping clingo from pre-built binaries
==> Warning: download_tarball() was unable to download clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64 ^python@3.6%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-x86_64 from any configured mirrors
==> Warning: Spec clingo-bootstrap@spack%gcc@9.3.0~docs~ipo+python build_type=Release arch=linux-rhel5-x86_64 ^python@3.6%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-rhel5-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@9.3.0~docs~ipo+python build_type=Release arch=linux-rhel5-x86_64 ^python@3.6%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-rhel5-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64 ^python@3.6%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-x86_64 did not provide module clingo
==> Warning: Spec clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64 ^python@3.6%gcc@10.2.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-x86_64 did not provide module clingo
==> Bootstrapping clingo from sources
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/libiconv-1.16-u3wai7xjn5pp67c362xdd5oco5u2w7s5
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/libsigsegv-2.13-sdf2wrllth6m6b7hmkf67odaagqked7w
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/berkeley-db-18.1.40-ajnefcynnzksacctdjmzfwavjc6rkrba
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/pkgconf-1.8.0-66p4neyoqpa4uoxd7w337kipcjcn7b23
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/zlib-1.2.12-346e5znx6mcfxp437utdu5kihfrjhym3
[+] /usr/local (external cmake-3.22.0-h5btcczbzvrvnupbubkfzz2c3ypgmsfy)
[+] /vol7/software/python3 (external python-3.6-qwfyel3i4lreszxtf4o3fto2syrqywed)
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/re2c-2.2-ygxjx3fxcogssfwjijn4pcybhnbdqdkh
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/diffutils-3.8-i7gdl4uh5j4waplvmupuv7fazwfg7dk7
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/ncurses-6.2-ozrz5qmebbqqpaklnahhd4dqrp3csvfh
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/bzip2-1.0.8-ukoq4ou2ytcwvn5xglkbsoodknba5va5
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/m4-1.4.19-af6l6v4gojang46ya4uekdjqswo5cr2y
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/readline-8.1.2-roykvwm4b7dddcxyluytdnuxjsreognw
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/gdbm-1.19-m4q7axferv6pmmdxese4cm3dnt642d4a
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/perl-5.34.1-zp7x2qjuneh4gznk4i3nt3cn4g2tcwqj
[+] /vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/bison-3.8.2-ugsp64z7myu5mjft6auevvvdo2k7m6xx
==> Installing clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn
==> No binary for clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn found: installing from source
==> Using cached archive: /vol7/home/lrtfm/opt/spack/var/spack/cache/_source-cache/git//potassco/clingo.git/2a025667090d71b2c9dce60fe924feb6bde8f667.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Already patched clingo-bootstrap
==> clingo-bootstrap: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/lib;/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/lib64;/vol7/software/python3/lib' '-DCMAKE_PREFIX_PATH:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/re2c-2.2-ygxjx3fxcogssfwjijn4pcybhnbdqdkh;/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/bison-3.8.2-ugsp64z7myu5mjft6auevvvdo2k7m6xx;/vol7/software/python3' '-DCLINGO_BUILD_WITH_LUA=OFF' '-DCLINGO_REQUIRE_PYTHON=ON' '-DCLINGO_BUILD_WITH_PYTHON=ON' '-DPYCLINGO_USER_INSTALL=OFF' '-DPYCLINGO_USE_INSTALL_PREFIX=ON' '-DCLINGO_BUILD_PY_SHARED:STRING=OFF' '-DPython_EXECUTABLE:STRING=/vol7/software/python3/bin/python3.6' '-DPython_INCLUDE_DIR:STRING=/vol7/software/python3/include/python3.6m' '-DCLINGO_BUILD_APPS:STRING=OFF' '/tmp/lrtfm/spack-stage/spack-stage-clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/spack-src'

1 error found in build log:
  >> 3    CMake Error at /usr/local/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
     4      Could NOT find Python (missing: Interpreter Development Development.Module
     5      Development.Embed) (found version "2.6")
     6    Call Stack (most recent call first):
     7      /usr/local/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
     8      /usr/local/share/cmake-3.22/Modules/FindPython.cmake:556 (find_package_handle_standard_args)
     9      CMakeLists.txt:130 (find_package)

See build log for details:
  /tmp/lrtfm/spack-stage/spack-stage-clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/spack-build-out.txt

==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
github-actions-v0.2 raised RuntimeError: download of binary cache file for spec "clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64" failed
github-actions-v0.1 raised ValueError: source is not trusted
spack-install raised InstallError: Terminating after first install failure: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/lib;/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/lib64;/vol7/software/python3/lib' '-DCMAKE_PREFIX_PATH:STRING=/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/re2c-2.2-ygxjx3fxcogssfwjijn4pcybhnbdqdkh;/vol7/home/lrtfm/.spack/bootstrap/store/linux-rhel6-x86_64/gcc-7.1.0/bison-3.8.2-ugsp64z7myu5mjft6auevvvdo2k7m6xx;/vol7/software/python3' '-DCLINGO_BUILD_WITH_LUA=OFF' '-DCLINGO_REQUIRE_PYTHON=ON' '-DCLINGO_BUILD_WITH_PYTHON=ON' '-DPYCLINGO_USER_INSTALL=OFF' '-DPYCLINGO_USE_INSTALL_PREFIX=ON' '-DCLINGO_BUILD_PY_SHARED:STRING=OFF' '-DPython_EXECUTABLE:STRING=/vol7/software/python3/bin/python3.6' '-DPython_INCLUDE_DIR:STRING=/vol7/software/python3/include/python3.6m' '-DCLINGO_BUILD_APPS:STRING=OFF' '/tmp/lrtfm/spack-stage/spack-stage-clingo-bootstrap-spack-c5kwvss4gadg42clnd2kddbrl6pmyonn/spack-src'
Run `spack --debug ...` for more detailed errors

</pre></details>

After I use `spack bootstrap trust` for `github-actions-v0.2` and `github-actions-v0.1`, the error for them changes to 
```
github-actions-v0.2 raised RuntimeError: The binary index is empty
github-actions-v0.1 raised RuntimeError: The binary index is empty
```
Other things are still the same.

### Information on your system

[lrtfm@ln0%tianhe2 spack-src]$ spack debug report
* **Spack:** 0.19.0.dev0
* **Python:** 3.6.0
* **Platform:** linux-rhel6-ivybridge
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/9181653/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/9181640/spack-build-out.txt)
[CMakeOutput.log](https://github.com/spack/spack/files/9181639/CMakeOutput.log)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by haampie** [comment]
If you're using system python you have to install python-dev too (with the headers)

**Comment by lrtfm** [comment]
Thanks. But it's not the system python, and the headers are all in `/vol7/software/python3/include/python3.6m`  which have been passed to cmake by `-DPython_INCLUDE_DIR:STRING=/vol7/software/python3/include/python3.6m`.

**Comment by lrtfm** [comment]
Now, I installed python myself, and the following CMake file can find python
```
cmake_minimum_required(VERSION 3.1)
find_package (Python COMPONENTS Interpreter Development)
```
with output
```
-- Found Python: /vol7/home/lrtfm/opt/python38/bin/python3.8 (found version "3.8.13") found components: Interpreter Development Development.Module Development.Embed
```

But when using spack to install the package, the CMake still can not find python's components "Development.Module: and "Development.Embed"

Is some environment variables need to be set?

**Comment by lrtfm** [comment]
First of all, the python is not installed by the system. Without `LD_LIBRARY_PATH`, it can not find the `so` file, which results in the failure of CMake. So I installed python by compiling it with `LDFLAGS=-Wl,-rpath=/where/I/will/install/python/lib`, 
and then add the python's binary path to `PATH`, then CMake found the new python.

However, another problem arises. As a new so `libstdc++.so.6` is installed in `/vol7/lib`, `re2c` and `patchelf` (which are need for building `clingo`) can not find this lib. But they were compiled without error.
This will also result in an error in configuring `clingo`. I added rpath `/vol7/lib` to `re2c` and `pathchelf`, then clingo build Successfully.

It seems that those bootstrap problems all relate to `LD_LIBRARY_PATH`. How can I make `LD_LIBRARY_PATH` work in the bootstrap stage?

In addition, I found that, in the bootstrap stage, `extra_rpath` and `flags`  in `compilers.yaml` seems not to work. Is this true? The option `dirty: true` in `config.yaml` seems not to work at this stage. Is this an expected behavior?
